File: backend/optimizer.py
from __future__ import annotations

from typing import Any

from config import COMPATIBILITY_FIT, MODE_TARGET_FILL
from utils import as_int


def route_anchor_node(label: str) -> str:
    primary = label.split("+", 1)[0].split("/", 1)[0].strip()
    return primary or label


def route_anchor_nodes(label: str) -> list[str]:
    nodes = []
    for part in label.replace("+", "/").split("/"):
        node = part.strip()
        if node and node not in nodes:
            nodes.append(node)
    return nodes or [label]


def route_nodes_for(corridor: dict[str, Any], shipment: dict[str, Any]) -> list[str]:
    corridor_origin = route_anchor_node(str(corridor["origin"]))
    corridor_origin_nodes = route_anchor_nodes(str(corridor["origin"]))
    corridor_destination = route_anchor_node(str(corridor["destination"]))
    corridor_destination_nodes = route_anchor_nodes(str(corridor["destination"]))
    mode = str(shipment["mode"])

    if mode == "air":
        nodes = [
            shipment["origin"],
            f"{corridor_destination} air cargo window",
            f"{corridor_origin} air gateway",
            shipment["destination"],
        ]
    elif mode == "sea":
        nodes = [
            shipment["origin"],
            f"{corridor_destination} feeder consolidation",
            f"{corridor_origin} port linkage",
            shipment["destination"],
        ]
    elif mode == "staging":
        nodes = [
            shipment["origin"],
            f"{corridor_destination} staging area",
            f"{corridor_origin} consolidation hub",
            shipment["destination"],
        ]
    else:
        nodes = [
            shipment["origin"],
            *corridor_destination_nodes,
            *reversed(corridor_origin_nodes),
            shipment["destination"],
        ]

    deduped: list[str] = []
    for node in nodes:
        node_text = str(node).strip()
        if node_text and node_text not in deduped:
            deduped.append(node_text)
    return deduped


def instruction_for(corridor: dict[str, Any], shipment: dict[str, Any], max_detour: int) -> str:
    mode = str(shipment["mode"])
    tonnes = as_int(shipment["matchedTonnes"])
    cargo = str(shipment["cargo"]).lower()

    if mode == "air":
        return (
            f"Book {tonnes} t of {cargo} into the next belly-cargo window and keep road movement "
            "limited to first and last mile transfers."
        )
    if mode == "sea":
        return (
            f"Batch {tonnes} t of {cargo} through the feeder or port linkage and use the corridor "
            "only for drayage and consolidation."
        )
    if mode == "staging":
        return (
            f"Stage {tonnes} t of {cargo} at the consolidation buffer, then release it with the "
            "next compatible dispatch wave."
        )
    return (
        f"Assign {tonnes} t of {cargo} to return truck capacity on {corridor['returnLane']} with "
        f"a detour cap of {max_detour} km."
    )


def build_recommended_actions(
    corridor: dict[str, Any],
    accepted: list[dict[str, Any]],
    available_by_mode: dict[str, int],
    max_detour: int,
    guardrail: int,
) -> list[dict[str, Any]]:
    actions = []
    for index, shipment in enumerate(accepted[:6], start=1):
        matched_tonnes = as_int(shipment["matchedTonnes"])
        mode = str(shipment["mode"])
        revenue = matched_tonnes * as_int(shipment["revenuePerTon"])
        empty_km = round(
            (matched_tonnes / (16 if mode == "road" else 24))
            * max(120, as_int(corridor["distanceKm"]) - as_int(shipment["detourKm"]))
        )
        available = max(available_by_mode.get(mode, 0), 1)
        capacity_share = round((matched_tonnes / available) * 100)
        route = route_nodes_for(corridor, shipment)

        actions.append(
            {
                "id": f"ACT-{index:02d}-{shipment['id']}",
                "shipmentId": shipment["id"],
                "headline": f"Move {matched_tonnes} t of {shipment['cargo']} by {mode}",
                "shipper": shipment["shipper"],
                "cargo": shipment["cargo"],
                "mode": mode,
                "matchedTonnes": matched_tonnes,
                "route": route,
                "routeText": " -> ".join(route),
                "operatingInstruction": instruction_for(corridor, shipment, max_detour),
                "why": [
                    f"Score {shipment['score']} from route fit, reliability, revenue and urgency",
                    f"{shipment['detourKm']} km detour is within the {max_detour} km policy",
                    f"{shipment['compatibility']} cargo clears the {guardrail}% compatibility guardrail",
                    f"{shipment['window']} service window with {shipment['reliability']}% reliability",
                ],
                "revenue": revenue,
                "emptyKmAvoided": empty_km,
                "capacityShare": capacity_share,
                "timing": f"Pickup {shipment.get('pickupWindow', shipment['window'])}; deliver by {shipment.get('deliveryWindow', shipment['window'])}",
            }
        )

    return actions


def score_shipment(shipment: dict[str, Any], max_detour: int, guardrail: int) -> int:
    route_fit = max(0, 1 - as_int(shipment["detourKm"]) / max(max_detour, 1))
    reliability_fit = as_int(shipment["reliability"]) / 100
    revenue_fit = min(as_int(shipment["revenuePerTon"]) / 12000, 1)
    urgency_fit = as_int(shipment["urgency"]) / 100
    compatibility_fit = COMPATIBILITY_FIT.get(str(shipment["compatibility"]), 0.5)
    guardrail_penalty = ((guardrail - 50) / 50) * (1 - compatibility_fit) * 22

    return round(
        route_fit * 32
        + reliability_fit * 24
        + compatibility_fit * 22
        + revenue_fit * 15
        + urgency_fit * 7
        - guardrail_penalty
    )


def optimise_corridor(
    corridor: dict[str, Any],
    anchor_enabled: bool,
    anchor_multiplier: float,
    max_detour: int,
    guardrail: int,
) -> dict[str, Any]:
    anchor_factor = anchor_multiplier if anchor_enabled else 0.54
    available_by_mode: dict[str, int] = {}
    remaining_bookable: dict[str, int] = {}

    for item in corridor["modes"]:
        available_capacity = round(as_int(item["capacity"]) * (as_int(item["available"]) / 100) * anchor_factor)
        target_fill = MODE_TARGET_FILL.get(str(item["mode"]), 0.86)
        # Bookable capacity is deliberately below available capacity to keep a real-world operating buffer.
        bookable_capacity = round(available_capacity * target_fill) if available_capacity else 0
        available_by_mode[item["mode"]] = available_capacity
        remaining_bookable[item["mode"]] = bookable_capacity

    scored = []
    for shipment in corridor["shipments"]:
        scored.append(
            {
                **shipment,
                "score": score_shipment(shipment, max_detour, guardrail),
                "matchedTonnes": 0,
                "reason": "",
            }
        )
    scored.sort(key=lambda item: item["score"], reverse=True)

    accepted: list[dict[str, Any]] = []
    declined: list[dict[str, Any]] = []
    threshold = max(55, guardrail - 12)

    for shipment in scored:
        mode = shipment["mode"]
        mode_remaining = remaining_bookable.get(mode, 0)
        incompatible = guardrail >= 82 and shipment["compatibility"] in {"chilled", "regulated"}
        detour_blocked = shipment["mode"] != "air" and as_int(shipment["detourKm"]) > max_detour

        if mode_remaining <= 0:
            declined.append({**shipment, "reason": "No capacity left in this mode"})
            continue

        if detour_blocked:
            declined.append({**shipment, "reason": "Detour exceeds lane policy"})
            continue

        if shipment["score"] < threshold or incompatible:
            declined.append({**shipment, "reason": "Guardrail confidence too low"})
            continue

        matched_tonnes = min(as_int(shipment["tonnes"]), mode_remaining)
        remaining_bookable[mode] = mode_remaining - matched_tonnes
        accepted.append(
            {
                **shipment,
                "matchedTonnes": matched_tonnes,
                "reason": "Full match"
                if matched_tonnes == as_int(shipment["tonnes"])
                else f"{matched_tonnes} t partial match",
            }
        )

    adjusted_capacity = sum(
        round(as_int(item["capacity"]) * (as_int(item["available"]) / 100) * anchor_factor)
        for item in corridor["modes"]
    )
    matched_tonnes = sum(as_int(item["matchedTonnes"]) for item in accepted)
    revenue = sum(as_int(item["matchedTonnes"]) * as_int(item["revenuePerTon"]) for item in accepted)
    empty_km_avoided = round(
        sum(
            (as_int(item["matchedTonnes"]) / (16 if item["mode"] == "road" else 24))
            * max(120, as_int(corridor["distanceKm"]) - as_int(item["detourKm"]))
            for item in accepted
        )
    )
    cost_saved = empty_km_avoided * 68
    load_factor = round((matched_tonnes / adjusted_capacity) * 100) if adjusted_capacity else 0
    baseline_load_factor = 19 if anchor_enabled else 11
    unit_cost_drop = max(0, min(34, round((load_factor - baseline_load_factor) * 0.62)))
    anchor_tonnes = round(as_int(corridor["baseAnchorTonnes"]) * anchor_factor)

    mode_utilisation = []
    total_remaining: dict[str, int] = {}
    for item in corridor["modes"]:
        mode = str(item["mode"])
        available = available_by_mode[mode]
        target_fill = MODE_TARGET_FILL.get(mode, 0.86)
        bookable_capacity = round(available * target_fill) if available else 0
        bookable_remaining = remaining_bookable.get(mode, 0)
        used = max(0, bookable_capacity - bookable_remaining)
        mode_remaining = max(0, available - used)
        total_remaining[mode] = mode_remaining
        utilisation = round((used / available) * 100) if available else 0
        mode_utilisation.append(
            {
                **item,
                "availableTonnes": available,
                "bookableTonnes": bookable_capacity,
                "remainingTonnes": mode_remaining,
                "reserveTonnes": max(0, available - bookable_capacity),
                "usedTonnes": used,
                "utilisation": utilisation,
            }
        )

    return {
        "accepted": accepted,
        "declined": declined,
        "recommendedActions": build_recommended_actions(
            corridor,
            accepted,
            available_by_mode,
            max_detour,
            guardrail,
        ),
        "remaining": total_remaining,
        "modeUtilisation": mode_utilisation,
        "adjustedCapacity": adjusted_capacity,
        "matchedTonnes": matched_tonnes,
        "revenue": revenue,
        "emptyKmAvoided": empty_km_avoided,
        "costSaved": cost_saved,
        "loadFactor": load_factor,
        "unitCostDrop": unit_cost_drop,
        "anchorTonnes": anchor_tonnes,
    }

File: app/lib/optimizer.ts
import { modeTargetFill } from "./config";
import type {
  Corridor,
  Mode,
  OptimizationResult,
  RecommendedAction,
  ScoredShipment,
  Shipment,
} from "./types";

function routeAnchorNode(label: string) {
  return label.split("+")[0].split("/")[0].trim() || label;
}

function routeAnchorNodes(label: string) {
  const nodes: string[] = [];

  label
    .replace(/\+/g, "/")
    .split("/")
    .forEach((part) => {
      const node = part.trim();
      if (node && !nodes.includes(node)) {
        nodes.push(node);
      }
    });

  return nodes.length ? nodes : [label];
}

function routeNodesFor(corridor: Corridor, shipment: ScoredShipment) {
  const corridorOrigin = routeAnchorNode(corridor.origin);
  const corridorOriginNodes = routeAnchorNodes(corridor.origin);
  const corridorDestination = routeAnchorNode(corridor.destination);
  const corridorDestinationNodes = routeAnchorNodes(corridor.destination);

  if (shipment.mode === "air") {
    return [
      shipment.origin,
      `${corridorDestination} air cargo window`,
      `${corridorOrigin} air gateway`,
      shipment.destination,
    ].filter((node, index, nodes) => node && nodes.indexOf(node) === index);
  }

  if (shipment.mode === "sea") {
    return [
      shipment.origin,
      `${corridorDestination} feeder consolidation`,
      `${corridorOrigin} port linkage`,
      shipment.destination,
    ].filter((node, index, nodes) => node && nodes.indexOf(node) === index);
  }

  if (shipment.mode === "staging") {
    return [
      shipment.origin,
      `${corridorDestination} staging area`,
      `${corridorOrigin} consolidation hub`,
      shipment.destination,
    ].filter((node, index, nodes) => node && nodes.indexOf(node) === index);
  }

  return [
    shipment.origin,
    ...corridorDestinationNodes,
    ...[...corridorOriginNodes].reverse(),
    shipment.destination,
  ].filter((node, index, nodes) => node && nodes.indexOf(node) === index);
}

function instructionFor(corridor: Corridor, shipment: ScoredShipment, maxDetour: number) {
  const cargo = shipment.cargo.toLowerCase();

  if (shipment.mode === "air") {
    return `Book ${shipment.matchedTonnes} t of ${cargo} into the next belly-cargo window and keep road movement limited to first and last mile transfers.`;
  }

  if (shipment.mode === "sea") {
    return `Batch ${shipment.matchedTonnes} t of ${cargo} through the feeder or port linkage and use the corridor only for drayage and consolidation.`;
  }

  if (shipment.mode === "staging") {
    return `Stage ${shipment.matchedTonnes} t of ${cargo} at the consolidation buffer, then release it with the next compatible dispatch wave.`;
  }

  return `Assign ${shipment.matchedTonnes} t of ${cargo} to return truck capacity on ${corridor.returnLane} with a detour cap of ${maxDetour} km.`;
}

function buildRecommendedActions(
  corridor: Corridor,
  accepted: ScoredShipment[],
  availableByMode: Map<Mode, number>,
  maxDetour: number,
  guardrail: number,
): RecommendedAction[] {
  return accepted.slice(0, 6).map((shipment, index) => {
    const route = routeNodesFor(corridor, shipment);
    const revenue = shipment.matchedTonnes * shipment.revenuePerTon;
    const emptyKmAvoided = Math.round(
      (shipment.matchedTonnes / (shipment.mode === "road" ? 16 : 24)) *
        Math.max(120, corridor.distanceKm - shipment.detourKm),
    );
    const available = Math.max(availableByMode.get(shipment.mode) ?? 0, 1);

    return {
      id: `ACT-${String(index + 1).padStart(2, "0")}-${shipment.id}`,
      shipmentId: shipment.id,
      headline: `Move ${shipment.matchedTonnes} t of ${shipment.cargo} by ${shipment.mode}`,
      shipper: shipment.shipper,
      cargo: shipment.cargo,
      mode: shipment.mode,
      matchedTonnes: shipment.matchedTonnes,
      route,
      routeText: route.join(" -> "),
      operatingInstruction: instructionFor(corridor, shipment, maxDetour),
      why: [
        `Score ${shipment.score} from route fit, reliability, revenue and urgency`,
        `${shipment.detourKm} km detour is within the ${maxDetour} km policy`,
        `${shipment.compatibility} cargo clears the ${guardrail}% compatibility guardrail`,
        `${shipment.window} service window with ${shipment.reliability}% reliability`,
      ],
      revenue,
      emptyKmAvoided,
      capacityShare: Math.round((shipment.matchedTonnes / available) * 100),
      timing: `Pickup ${shipment.pickupWindow ?? shipment.window}; deliver by ${
        shipment.deliveryWindow ?? shipment.window
      }`,
    };
  });
}

function scoreShipment(shipment: Shipment, maxDetour: number, guardrail: number) {
  const routeFit = Math.max(0, 1 - shipment.detourKm / Math.max(maxDetour, 1));
  const reliabilityFit = shipment.reliability / 100;
  const revenueFit = Math.min(shipment.revenuePerTon / 12000, 1);
  const urgencyFit = shipment.urgency / 100;
  const compatibilityFit =
    shipment.compatibility === "dry" || shipment.compatibility === "ambient"
      ? 1
      : shipment.compatibility === "fragile"
        ? 0.86
        : shipment.compatibility === "regulated"
          ? 0.74
          : 0.42;
  const guardrailPenalty = ((guardrail - 50) / 50) * (1 - compatibilityFit) * 22;

  return Math.round(
    routeFit * 32 +
      reliabilityFit * 24 +
      compatibilityFit * 22 +
      revenueFit * 15 +
      urgencyFit * 7 -
      guardrailPenalty,
  );
}

export function optimizeCorridor(
  corridor: Corridor,
  anchorEnabled: boolean,
  anchorMultiplier: number,
  maxDetour: number,
  guardrail: number,
): OptimizationResult {
  const anchorFactor = anchorEnabled ? anchorMultiplier : 0.54;
  const availableByMode = new Map<Mode, number>();
  const remainingBookable = new Map<Mode, number>();

  corridor.modes.forEach((item) => {
    const availableCapacity = Math.round(item.capacity * (item.available / 100) * anchorFactor);
    // The fallback model mirrors the backend: keep operating reserve so bars do not unrealistically peg at 100%.
    const bookableCapacity = availableCapacity ? Math.round(availableCapacity * modeTargetFill[item.mode]) : 0;
    availableByMode.set(item.mode, availableCapacity);
    remainingBookable.set(item.mode, bookableCapacity);
  });

  const scored = corridor.shipments
    .map((shipment) => ({
      ...shipment,
      score: scoreShipment(shipment, maxDetour, guardrail),
      matchedTonnes: 0,
      reason: "",
    }))
    .sort((a, b) => b.score - a.score);

  const accepted: ScoredShipment[] = [];
  const declined: ScoredShipment[] = [];
  const threshold = Math.max(55, guardrail - 12);

  scored.forEach((shipment) => {
    const modeRemaining = remainingBookable.get(shipment.mode) ?? 0;
    const incompatible =
      guardrail >= 82 &&
      (shipment.compatibility === "chilled" || shipment.compatibility === "regulated");
    const detourBlocked = shipment.mode !== "air" && shipment.detourKm > maxDetour;

    if (modeRemaining <= 0) {
      declined.push({ ...shipment, reason: "No capacity left in this mode" });
      return;
    }

    if (detourBlocked) {
      declined.push({ ...shipment, reason: "Detour exceeds lane policy" });
      return;
    }

    if (shipment.score < threshold || incompatible) {
      declined.push({ ...shipment, reason: "Guardrail confidence too low" });
      return;
    }

    const matchedTonnes = Math.min(shipment.tonnes, modeRemaining);
    remainingBookable.set(shipment.mode, modeRemaining - matchedTonnes);
    accepted.push({
      ...shipment,
      matchedTonnes,
      reason: matchedTonnes === shipment.tonnes ? "Full match" : `${matchedTonnes} t partial match`,
    });
  });

  const adjustedCapacity = corridor.modes.reduce((sum, item) => {
    return sum + Math.round(item.capacity * (item.available / 100) * anchorFactor);
  }, 0);
  const matchedTonnes = accepted.reduce((sum, item) => sum + item.matchedTonnes, 0);
  const revenue = accepted.reduce((sum, item) => sum + item.matchedTonnes * item.revenuePerTon, 0);
  const emptyKmAvoided = Math.round(
    accepted.reduce((sum, item) => {
      const distance = Math.max(120, corridor.distanceKm - item.detourKm);
      const vehicleEquivalent = item.matchedTonnes / (item.mode === "road" ? 16 : 24);
      return sum + vehicleEquivalent * distance;
    }, 0),
  );
  const costSaved = emptyKmAvoided * 68;
  const loadFactor = adjustedCapacity ? Math.round((matchedTonnes / adjustedCapacity) * 100) : 0;
  const baselineLoadFactor = anchorEnabled ? 19 : 11;
  const unitCostDrop = Math.max(0, Math.min(34, Math.round((loadFactor - baselineLoadFactor) * 0.62)));
  const anchorTonnes = Math.round(corridor.baseAnchorTonnes * anchorFactor);
  const totalRemaining = new Map<Mode, number>();

  corridor.modes.forEach((item) => {
    const availableCapacity = availableByMode.get(item.mode) ?? 0;
    const bookableCapacity = availableCapacity ? Math.round(availableCapacity * modeTargetFill[item.mode]) : 0;
    const bookableRemaining = remainingBookable.get(item.mode) ?? 0;
    const usedCapacity = Math.max(0, bookableCapacity - bookableRemaining);
    totalRemaining.set(item.mode, Math.max(0, availableCapacity - usedCapacity));
  });

  return {
    accepted,
    declined,
    recommendedActions: buildRecommendedActions(
      corridor,
      accepted,
      availableByMode,
      maxDetour,
      guardrail,
    ),
    remaining: Object.fromEntries(totalRemaining) as Partial<Record<Mode, number>>,
    adjustedCapacity,
    matchedTonnes,
    revenue,
    emptyKmAvoided,
    costSaved,
    loadFactor,
    unitCostDrop,
    anchorTonnes,
  };
}
